text/trace/115-add-sample-to-span-end.md

```diff
@@ -0,0 +1,182 @@
+  payload, after some bytes have been read.
+
+The span of the operation should have a start time that is right before the
+first byte is received.
+
+### Deferred span creation is worse than creating a span
+
+Ensuring a span is properly closed, even in error scenarios, is a desirable
+trait, and SDKs have ensured best-effort of closing of spans using language
+specific features such as exception handling or deferred close routines.
+
+Deferred span creation has not had as much development focus, which results in
+a less error handling that ensures that deferred spans are being created. This
+results in exceptional situations that occur before the span creation (e.g.
+invalid parsing of an http request) to not be represented in a span, as the
+deferred span is never created.
+
+A way to mitigate this would be to ensure the same level of error handling
+for deferred span creation as with spans themselves (create the span during a
+handled exception). But that would double the amount of maintenance to ensure
+robust span creation. Without span sampling as a concern, it would be desirable
+to just use a span instead, and focus on improving the robustness span creation.
+
+There is another side effect of deferred span creation, where there may be
+child spans that are created before the deferred span is. Examples here include
+a span to represent the receiving of data: the span that represents receiving data
+would have started before the name of the span is available, since that depends
+on receiving and parsing the incoming payload. The result will at worst be
+multiple root spans that should actually be children of the deferred span, since
+there will be no active span present when the children are created.
+
+### It is better to address the issues with sampling
+
+Given the need for an almost identical, most likely less robust code path
+for deferred spans, and the fact that it is a very common pattern, solving
+issues with sampling would ensure a more robust and maintainable code base.
+
+## Explanation
+
+A sampler would consume the same arguments it does today:
+
+- span name
+- span context
+- span kind
+- span attributes
+- span links
+
+The sampler would be called twice: once at creation, and once at span end.
+
+The span's sampling decision would be determined by the intersection of the
+sampling decision made by either ShouldSample call:
+
+| First ShouldSample result | Second ShouldSample result | Final Result       |
+|---------------------------|----------------------------|--------------------|
+| NOT_RECORD                | N/A (Don't Call)           | NOT_RECORD         |
+| RECORD_AND_SAMPLED        | N/A (Don't Call)           | RECORD_AND_SAMPLED |
+| RECORD                    | NOT_RECORD                 | NOT_RECORD         |
+| RECORD                    | RECORD                     | RECORD             |
+| RECORD                    | RECORD_AND_SAMPLED         | RECORD_AND_SAMPLED |
```

Comment:
I would actually go the other way (as we do in Jaeger clients): No can be upgraded to Yes, but Yes cannot be downgraded.

Reply:
I think that works in most cases, with the exception of NOT_RECORD. 

As I understand it, some implementations are taking a NOT_RECORD initial decision as a way to insert in a noop span, which will not be able to record further values or changes. So that is irreversible:

https://github.com/open-telemetry/opentelemetry-specification/issues/530#issuecomment-605980247

Assuming that constraint, the only thing that changes in the table is that RECORD + NOT_RECORD => RECORD

Reply:
> As I understand it, some implementations are taking a NOT_RECORD initial decision as a way to insert in a noop span

That's fine in the implementation does not support up-sampling at all, as in the most basic implementation of upfront sampling where the first decision is permanent. If they want to support _any form of delayed sampling_, noop span simply doesn't make sense.

Allowing partial downsampling (excluding some subtree from the trace) makes it pretty difficult to correctly interpret effective sampling probability when deriving statistics from sampled data.

Reply:
I thought one of the primary goals of the sampling API was to support attaching a noop span some of the time for users that do not want to cost of delayed sampling. What I see here works as intended.

Reply:
the mechanism in Jaeger that I described does not prevent the use of noop, because of the `retryable` flag.

Reply:
It sounds like we should be thinking of the first decision as 3-valued (yes, no, maybe), where "maybe" corresponds with a retryable No decision in Jaeger?

I support this proposal and I think it will resolve https://github.com/open-telemetry/opentelemetry-specification/issues/620 too.

Reply:
yes, with the caveat that `retryable` in Jaeger is local to the process, Jaeger does not support "maybe" between processes like Zipkin does (although without reverse propagation I don't know what benefit Zipkin gets from "maybe"). So if we want to support "maybe" between processes, it's a different design discussion from what `retryable` provides in Jaeger samplers.

Reply:
Great. I was not considering cross-process coordination in this way. A good solution that problem would be quite a sight.

Reply:
Just an FYI I've read through this thread and trying to gather a few thoughts around it.

Reply:
So I think that, as has been alluded, there is no robust solution for ensuring that a mutating samplling decision can be honored in all child spans, propagated or not. At least with this type of sampling. Aside from samplers who make decisions on variables are independent of the contents of the span, there will be spans that will not have all attributes until right when the span should end, which in turn means that samplers could be written that could modify sampling decisions right up until the end of the span, well after all children have been created.

A good example I can come up with is a sampler that leverages HTTP status code: that will not be known until after all application code, which means that a modified sampling decision when that attribute is set will be well after all remote requests.

If one wants to enable significant flexibility to the sampling logic, I think that has to be handled at the tail, which has the ability to see all spans and attributes in a trace, and make a decision from that. 

So anything we do here is best-effort. And I do believe there's a class of samplers that can do an almost perfect job: ones that depend on attributes that are set before business logic. For example, samplers that are dependent on http method + http path.

With all that said, I think I'm going to change this proposal to be very similar to https://github.com/open-telemetry/opentelemetry-specification/issues/307, effectively enabling updated decisions when any data around the span has been modified. And I'll clarify that this cannot ensure consistent sampling decisions across all spans in a trace, as is the case today.

> It sounds like we should be thinking of the first decision as 3-valued (yes, no, maybe), where "maybe" corresponds with a retryable No decision in Jaeger?

I'm going to do some thinking around the most appropriate semantics, but I do believe something that states a tentative result (needed for propagation) along with some ability to state whether the decision should be re-evaluated, is needed.

Reply:
I think we could add a little detail about sampling probability -- in the case where an in-band message is sent in one of these speculative ("maybe") states, you may only know an upper bound on the sampling probability. For example, a head sampler is configured with 10% sampling and a tail sampler is configured to further adjust the final probabilities. The in-band sampling probability (0.1) is an upper bound.  Sampling probability can be adjusted down (toward greater "sample count") by a later decision, but the probability cannot be adjusted up (toward lower "sample count").

Reply:
Can you clarify the value of the sampling probability? Is that to provide information for future samplers to make a more informed decision?

That feels like a valuable component (and something #107) might address, but unrelated to the introduction of a "maybe" state. I think to enable this OTEP of multiple sampling decision points, only the "maybe" state is needed.

Reply:
sorry to be clear I think it's a good idea and happy to write another OTEP after this one (or support an existing one), but want to keep scope specific to the problem to help drive a decision.

Reply:
It's fine to leave this detail out--I wasn't trying to derail your progress. I was anticipating that someone would raise an objection that we don't know what value to propagate for sampling probability when the in the "maybe" state, but I think we can answer that question. The value of propagating sampling probability is to support some kind of rate limiting in the downstream service.

Reply:
yes, makes sense!

I've updated the document to advocate for @yurishkuro's suggestion of a retryable value instead (which I've named shouldRetry based on the expected behavior, but that's up for debate).

The rationale was due to the concern about what "MAYBE" should be. I view there being design tradeoffs with the two choices:

* MAYBE -> propagate yes results in more spans than desired sent to the backend. This is a good idea for systems that favor completeness over the span consumer's resource utilization.
* MAYBE -> propagate no results in fewer spans sent to the backend, effectively missing spans. This is better for systems that favor resource utilization over completeness.

As such, I would say it's better to leave that decision as an exercise for the span processor, and have the SDK enable both options.